doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+  generators need to be explicitly created within the :func:`@qjit <~.qjit>` function
+  using ``jax.random.PRNGKey(int)``:
+
+  >>> @qjit()
+  ... def f():
+  ...     key = jax.random.PRNGKey(0)
+  ...     a = jax.random.normal(key, shape=(1,))
+  ...     return a
+  >>> f()
+  array([-0.78476578])
+
+* **Dynamic-shaped arrays:** Functions that create or return arrays with
+  dynamic shape --- that is, arrays where their shape is determined by a
+  dynamic variable at runtime -- are currently not supported in JAX nor
+  Catalyst. Typically, workarounds involve rewriting the code to utilize
+  ``jnp.where`` where possible.
+
+For more details, please see the `JAX documentation
+<https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html>`__.
+
+JAX integration
+---------------
+
+Compiled functions remain JAX compatible, and you can call JAX transformations
+on them, such as ``jax.grad`` and ``jax.vmap``. You can even call ``jax.jit``
+on functions that call qjit-compiled functions:
+
+>>> dev = qml.device("lightning.qubit", wires=2)
+>>> @qjit
+... @qml.qnode(dev)
+... def circuit(x):
+...     qml.RX(x, wires=0)
+...     return qml.expval(qml.PauliZ(0))
+>>> @jax.jit
+... def workflow(y):
+...     return jax.grad(circuit)(jnp.sin(y))
+>>> workflow(0.6)
+Array(-0.53511382, dtype=float64, weak_type=True)
+
+However, a ``jax.jit`` function calling a ``qjit`` function will always result
+in a callback to Python, so will be slower than if the function was purely compiled
+using ``jax.jit`` or ``qjit``.
+
+If you want to compile some functionality that is not currently Catalyst
+compatible, or you want to make use of JAX-supported hardware such as TPUs
+for classical processing, mixing ``jax.jit`` and ``qjit`` will allow this.
+However, if possible, try to always use ``qjit`` to compile your entire
+workflow.
+
+Internal QJIT transformations
+-----------------------------
+
+Inside of a qjit-compiled function, JAX transformations
+(``jax.grad``, ``jax.jacobian``, ``jax.vmap``, etc.)
+can be used **as long as they are not applied to quantum processing**.
+
+>>> @qjit
+... def f(x):
+...     def g(y):
+...         return -jnp.sin(y) ** 2
+...     return jax.grad(g)(x)
+>>> f(0.4)
+array(-0.71735609)
+
+If they are applied to quantum processing, an error will occur:
+
+>>> @qjit
+... def f(x):
+...     @qml.qnode(dev)
+...     def g(y):
+...         qml.RX(y, wires=0)
+...         return qml.expval(qml.PauliX(0))
+...     return jax.grad(lambda y: g(y) ** 2)(x)
+>>> f(0.4)
+NotImplementedError: must override
+
+Instead, only Catalyst transformations will work when applied to hybrid
+quantum-classical processing:
+
+>>> @qjit
+... def f(x):
+...     @qml.qnode(dev)
+...     def g(y):
+...         qml.RX(y, wires=0)
+...         return qml.expval(qml.PauliZ(0))
+...     return grad(lambda y: g(y) ** 2)(x)
+>>> f(0.4)
+array(-0.71735609)
+
+Always use the equivalent Catalyst transformation
```

Comment:
Do we have a performance comparison?

Reply:
We don't, no! Interestingly, I imagine `jax.grad` might be slower, since the transformation happens in Python/jaxpr, while `catalyst.grad` the transformation happens in MLIR 🤔

Reply:
Compilation a bit slower:

![image](https://github.com/PennyLaneAI/catalyst/assets/2959003/58a447ea-0fc9-4d61-9288-52623738016a)

But resulting binary approximately no difference:

![image](https://github.com/PennyLaneAI/catalyst/assets/2959003/f48e53ef-57ec-49e3-b554-0edee951e931)

Reply:
Very cool! The "differentiation process" in Python need not necessarily be slower since it happens at a very high level, Enzyme has a lot more instructions to process.

That the execution speed is slightly slower with Enzyme is interesting, it would be interesting to investigate how this shakes out on larger examples.
I remember asking people at the Enzyme conference how the performance compares to a framework like JAX, but the answers weren't so clear. Enzyme proposes that differentiation _can_ produce faster code if you perform it after code optimizations, but in practice it might be difficult to beat high-level AD.

Reply:
@dime10 yeah would be interesting to share this with the enzyme team :)

Reply:
>That the execution speed is slightly slower with Enzyme is interesting, it would be interesting to investigate how this shakes out on larger examples.

I think I tested some larger examples after I posted this comment, and it did in fact seem to scale worse. E.g., take the example above but execute with a 100000 sized array, and enzyme was ~2-5x slower iirc